From: Ratchanan Srirattanamet <ratchanan@ubports.com>
Date: Wed, 9 Oct 2019 00:06:05 +0700
Subject: [nemo-qtmultimedia-plugins] make sure backend's nativeSize isn't
 affected by orientation settings

QML VideoOutput expects backend's nativeSize to be unaffected by
orientation settings. This causes VideoOutput's contentRect and the
calculation of normalized point to be incorrect. This commit make
m_nativeSize be affected by texture orientation only and make
updatePaintNode() aware of this fact.

Forwarded: https://git.sailfishos.org/mer-core/nemo-qtmultimedia-plugins/merge_requests/6
Last-Update: 2019-10-10
diff --git a/src/videotexturebackend/videotexturebackend.cpp b/src/videotexturebackend/videotexturebackend.cpp
index 9f61de3..0dd0fcf 100644
--- a/src/videotexturebackend/videotexturebackend.cpp
+++ b/src/videotexturebackend/videotexturebackend.cpp
@@ -717,7 +717,12 @@ QSGNode *NemoVideoTextureBackend::updatePaintNode(QSGNode *oldNode, QQuickItem::
     if (m_geometryChanged) {
         const QRectF br = q->boundingRect();
 
-        QRectF rect(QPointF(0, 0), QSizeF(m_nativeSize).scaled(br.size(), Qt::KeepAspectRatio));
+        QSizeF size(m_nativeSize);
+        if ((m_orientation % 180) != 0)
+            size.transpose();
+        size.scale(br.size(), Qt::KeepAspectRatio);
+
+        QRectF rect(QPointF(0, 0), size);
         rect.moveCenter(br.center());
 
         int orientation = (m_orientation - m_textureOrientation) % 360;
@@ -798,12 +803,16 @@ bool NemoVideoTextureBackend::event(QEvent *event)
             {
                 QMutexLocker locker(&m_mutex);
                 m_nativeSize = nativeSize;
-                if ((m_orientation % 180) != 0) {
-                    m_nativeSize.transpose();
-                }
+            }
+
+            // FIXME: isn't this done by QDeclarativeVideoOutput already?
+            // https://github.com/qt/qtmultimedia/blob/7f5d08636ca9278e4a018db39b701457ea1e7115/src/qtmultimediaquicktools/qdeclarativevideooutput.cpp#L372
+            QSize itemImplicitSize = nativeSize;
+            if ((m_orientation % 180) != 0) {
+                itemImplicitSize.transpose();
             }
             static_cast<ImplicitSizeVideoOutput *>(q)->setImplicitSize(
-                        nativeSize.width(), nativeSize.height());
+                        itemImplicitSize.width(), itemImplicitSize.height());
         }
         q->update();
         emit nativeSizeChanged();
-- 
2.17.1

